Add ability to cycle GOCART aerosols

WalterKolczynski-NOAA opened the pull request:
The ability to cycle GOCART aerosols for otherwise cold-start initial conditions is added to the CROW-based coupled workflow.

The following now happen automatically when GOCART coupling is on
- The restart interval is modified to ensure one is created at the forecast cadence frequency in addition to any other times that may be specified.
- A new job, init_chem, is run on cycles after the first that copies the aerosol concentration from the previous cycle's restart files and  adds it to the cold-start initial conditions.

Importantly, the aerosol fields are merely copied to the initial condition files. **They are not modified for mass conservation or any other properties.**

As part of this modification, the method for specifying the forecast cadence was overhauled. gfs_cyc has been eliminated in favor of an explicit specification of the step size for both the GFS and DA system via STEP_GFS and STEP_DA, respectively. This allows for steps that are longer than 24 hours (See #260). Technically, the step sizes can be any arbitrary length, but other downstream settings are only listed in hours, so fractional hours may not work properly.

Setting STEP_GFS to !timedelta "00:00:00" will turn off GFS forecasts (DA cycling-only)
Setting STEP_DA to !timedelta "00:00:00" will turn off DA (free-forecast mode)

Note that setting STEP_DA to zero is now required for free forecasts in addition to using the free forecast CROW layout.

Examples are available in the commit message for c196176.

Refs: #260 
Closes: #314

WalterKolczynski-NOAA committed:
Improve cycle cadence specification in CROW

Replaces the old gfs_cyc method for determining cycling frequency with one that
explicitly sets the step length. In place of gfs_cyc there are two new variables
to determine step sizing: STEP_GFS and STEP_DA. Both are of the timedelta type.

STEP_GFS determines the frequency of GFS and defaults to 6 hours
STEP_DA determines the frequency of DA and also defaults to 6 hours

If STEP_GFS is set to 00:00:00, only the DA system is run
If STEP_DA is set to 00:00:00, only the GFS system is run (free-forecast mode)

Note that this new method currently requires the additional setting of STEP_DA
to zero, in addition to using free forecast layout.

If STEP_DA is non-zero, the first cycle is DA-only, so this should be kept in
mind when setting up your experiments (especially if you want GFS to run at
certain times of day). If both are non-zero, STEP_GFS should always be an
integer multiple of STEP_DA, otherwise unexpected behavior may occur. Other
than setting STEP_DA to zero for free forecast, STEP_DA is unlikely to be
changed from its default value in the near-future for GFS, but the new method
is flexible for any future incorporation of LAM system that requires faster DA
cycling.

Examples:

Full-cycling with DA & GFS:
STEP_GFS: !timedelta "06:00:00"
STEP_DA: !timedelta "06:00:00"
(since these are defaults, nothing needs to be placed in the case file, but you may
wish to for clarity)

Normal DA with GFS at each 00z starting 2013-04-01:
SDATE: 2013-03-31t18:00:00
STEP_GFS: !timedelta "24:00:00"

Free forecast with GFS every 24h at 00z starting 2013-04-01:
SDATE: 2013-04-01t00:00:00
STEP_GFS: !timedelta "24:00:00"
STEP_DA: !timedelta "00:00:00"

CROW ss updated to add a new tool that converts a timedelta to a string.

The wave model, which formerly used gfs_cyc to determine the wave cycling interval,
now determines WAVHCYC directly from STEP_GFS. Since this variable is only a number
of hours, this will only work for whole-hour GFS frequencies.

Case files have been updated to use the new settings in place of gfs_cyc, as well as
set STEP_DA to zero for free forecast cases. However, they have not all been tested.

Additional testing is needed to ensure these changes are fully operational for DA
cycling runs.

Refs: #260, #314

WalterKolczynski-NOAA committed:
Add restart output for aerosol cycling

Updates the restart_interval setting to make sure it always includes the
STEP_GFS time if GOCART is on. This is done using the advanced functionality
of restart_interval, which allows for an explicit list of restart times
instead of a single interval.

If STEP_GFS is a multiple of restart_interval, it is left unchanged.
If STEP_GFS is not a multiple of restart_interval, restart_interval is
converted to an explicit sequence of restart times, then STEP_GFS is added.
If restart_interval is already an explicit list, STEP_GFS is just added to it.
This may add a redundant restart time, but that is ignored by FV3.

In order to complete this update, the CROW type for restart_interval had to be
changed from int to string to accomodate lists of numbers. This has the added
benefit of allowing explicit definition of restart times in other applications
as well.

As part of this update, the values for restart_interval in the forecast config
were updated to use the appropriate settings rather than hardcoded values.
The default value for the gfs restart_interval was also changed from 6 to 0 to
maintain the previous behavior when the setting was ignored.

Refs: #314

WalterKolczynski-NOAA committed:
Add init_chem job to coupled workflow

A new init_chem job is added to the coupled workflow to cycle tracer
concentrations from the previous cycle's restart files. The new job
doesn't run for the first cycle (since there are no previous cycle
restart files to use). If aerosol initialization is desired for the
first cycle, they need to be added manually, though the new scripts
may be reutilized to do so.

Tracer concentrations are merely copied from the restart files to the
cold-start initial conditions. There is no adjustment to ensure mass
conservation or any other property. The list of tracers copied is
determined by the new parm/chem/gocart_tracer.list file, which is a
plain-text list of NetCDF variables to copy. If any are missing in the
restart files, the job will fail.

The NetCDF4 module for python3 is required for the init job to work.
This module is not part of the default installation of python on Hera
(though it is on WCOSS), so in addition to now loading python, the
base modulefile also adds the location of a personal copy of the
needed python module to the PYTHONPATH.

Most of the new scripts are recycled from the GEFS-aerosol development,
though this method was abandoned there in favor of true warm start for
the aerosol member.

Refs: #314

WalterKolczynski-NOAA commented:
A couple things that I hope can be updated before merging:
- Currently using a custom installation of hpc-stack to use the NASA version of a module. Hope to move to an official installation.
- Want to eliminate the separate checkout version and repo for the GOCART-enabled UFS.

WalterKolczynski-NOAA commented:
@rmontuoro in particular needs to confirm these changes are doing what he expected.

KateFriedman-NOAA commented on workflow/cases/dev/fv3q2fy19retro5-GFS@C768_ENKF@384+80MEM_whatever.yaml:
```diff
@@ -186,5 +186,5 @@ case:
     chgres_and_convert_ics: No
     realtime: false
-    gfs_cyc: 4 # run gfs at 0z, 12z
+    STEP_GFS: !timedelta "06:00:00"
```
What does STEP_DA default to if it's not set in the case file?

WalterKolczynski-NOAA replied:
STEP_DA defaults to !timedelta "06:00:00"

KateFriedman-NOAA replied:
Gotcha, thanks!

KateFriedman-NOAA commented on workflow/config/fcst.yaml:
```diff
@@ -231,6 +251,7 @@ config_fcst:
         export CHM_CFGDIR="{doc.chem_settings.CFGDIR}"
         export CHM_INPDIR="{doc.chem_settings.INPDIR}"
         export dnats="$(( $dnats + {doc.chem_settings.ntdiag} ))"
+        export KMP_INIT_AT_FORK=FALSE # Temporary setting
```
Curious about this variable and how it helps...this looks like something I considered for Comet.

WalterKolczynski-NOAA replied:
This was suggested by Raffaele for multithread problems in the aerosol model, though I don't know that it helped.

KateFriedman-NOAA commented on workflow/config/wave.yaml:
```diff
@@ -78,12 +78,8 @@ config_wave:
-        WAVHCYC=`expr 24 / $gfs_cyc`
       else
-        WAVHCYC=0
+        WAVHCYC={tools.dt_to_HMS(doc.settings.STEP_GFS).split(':')[0]}
```
Nice!

KateFriedman-NOAA commented on workflow/layout/cycled_gfs.yaml:
```diff
@@ -731,7 +731,7 @@ suite: !Cycle
     jgfs_archive: !Task
       <<: *service_task_template
-      Disable: !calc doc.settings.gfs_cyc == 0
+      Disable: !calc doc.settings.STEP_GFS == doc.zero_dt
```
Curious what zero_dt is here.

KateFriedman-NOAA commented on workflow/runtime/suite.yaml:
```diff
@@ -2,6 +2,9 @@
 # specific to the cycled workflow which is not in this release.
 
+day_dt: !timedelta "24:00:00"
+zero_dt: !timedelta "00:00:00"
```
I think this answers my question earlier...these are constants (0 and 24) that you use elsewhere?

WalterKolczynski-NOAA replied:
Yes

KateFriedman-NOAA commented on workflow/runtime/suite.yaml:
```diff
@@ -39,31 +42,41 @@ cycled_suite_alarms: &cycled_suite_alarms
     end: !calc doc.settings.EDATE
-    step: !timedelta "6:00:00"
+    step: !FirstTrue
+      - when: !calc doc.settings.STEP_DA == doc.zero_dt
```
Is this where the default for STEP_DA is set?

WalterKolczynski-NOAA replied:
Default is !timedelta 06:00:00

WalterKolczynski-NOAA replied:
No, that is set in defaults/settings.yaml. This is determining the master clock for the experiment. Since DA can happen more often than the forecast, STEP_DA is used unless it is off (zero).

KateFriedman-NOAA reviewed (approved):
I'm here just to see the revamp of gfs_cyc and related workflow updates. I like!

JessicaMeixner-NOAA commented on jobs/rocoto/init_chem.sh:
```diff
@@ -0,0 +1,56 @@
+status=$?
+[[ $status -ne 0 ]] && exit $status
+
+# export DATAROOT="$RUNDIR/$CDATE/$CDUMP"
```
delete commented out lines?

JessicaMeixner-NOAA commented on modulefiles/module_base.hera:
```diff
@@ -4,7 +4,8 @@
 # Modules availble through hpc-stack
-module use /scratch2/NCEPDEV/nwprod/hpc-stack/libs/hpc-stack/modulefiles/stack
+# module use /scratch2/NCEPDEV/nwprod/hpc-stack/libs/hpc-stack/modulefiles/stack
+module use /scratch1/NCEPDEV/nems/Raffaele.Montuoro/dev/nasa/libs/hpc-stack/1.1.0-nasa/modulefiles/stack
```
What's the path forward with not relying on Raffaele's stack? What's the timeline for getting what's needed in the emc hpc stack?

JessicaMeixner-NOAA replied:
These are also only for hera, so if we are doing this, shouldn't we be consistent throughout the platforms.

JessicaMeixner-NOAA replied:
Also in the strategy used here: https://github.com/ufs-community/ufs-weather-model/pull/591 only the modules that are different use the personal stack, perhaps that's a better way forward?  So only: 
module use -a /scratch1/NCEPDEV/nems/Raffaele.Montuoro/dev/nasa/ufs/lib/modulefiles

 module load gftl-shared/v1.3.0
 module load yafyaml/v0.5.1
 module load mapl/v2.7.1-beta

 module load intelpython/3.6.8

While all others use the hpc-stack from nwprod?

JessicaMeixner-NOAA commented on modulefiles/module_base.hera:
```diff
@@ -32,11 +34,14 @@ module load hdf5/1.10.6
 module load ncl/6.5.0
 module load cdo/1.9.5
+module load intelpython/3.6.8
+# netCDF4 is not in the python installation on Hera
```
So same thing here, we're moving away from hpc-stack or official modules.  Is this what we want to do?

WalterKolczynski-NOAA replied:
As noted in the commit message, this is definitely a work-around because NetCDF4 is not in the default python build on Hera (it is on WCOSS). We can talk to RDHPCS about having it added, though often their solution is "make your own conda environment", which is pretty much the same problem.

Personally I'd like to have a full python installation/environment in hpc-stack so it is standardized on all machines, but that path is something for @arunchawla-NOAA to determine. Alternatively, just the pyNetCDF4 module could be added there instead of my personal directory.

JessicaMeixner-NOAA reviewed (commented):
I have a couple of comments on the module files as maybe a different path forward.  Otherwise, as long as the ATM-OCN-ICE-WAV test passes, I don't see anything that would conflict with that application so should be okay.

WalterKolczynski-NOAA committed:
Fix print statement in chem init

The print statement for the tracer (restart) file search was missing
an f, so variables were not being substituted.

Refs: #314

WalterKolczynski-NOAA committed:
Fix minor issue with restart interval

The restart_interval export needed to be enclosed in quotation marks
so bash will properly assign the whole list to the variable.

Refs: #314

WalterKolczynski-NOAA committed:
Correct cycling interval issues for aerosols

FHCYC was being used to determine the forecast interval for the aerosol
cycling, but this appears to be an appropriate setting to use. Added a
new exported environment variable STEP_GFS, which is the number of hours
between GFS forecasts. While it is derived from the workflow STEP_GFS
setting, it is only the integer number of hours, not a timedelta like
the workflow setting. May want to revist this later to avoid confusion.

Additionally, the restart file written at the end of the forecast does
not have the timestamp as part of the filename like the intermediary
restart files do. The chem init script is adjusted to not use the time-
stamp when looking for the final restart file. And, since it now has to
read in FHMAX anyway, also added a short-circuit so the search doesn't
bother trying to search previous cycles that can't include the restart
time needed.

Refs: #314

rmontuoro reviewed (approved):
I've confirmed that these changes allow to successfully perform cycled runs with aerosols.
